src/utils/lazy.ts

```diff
@@ -7,21 +7,68 @@ export class Lazy<T> {
     private _isValueCreated: boolean = false;
     private _value: T | undefined;
 
-    public constructor(private readonly valueFactory: () => T) {
+    public constructor(private readonly valueFactory: () => T, private readonly _valueLifetime?: number) {
     }
 
     public get isValueCreated(): boolean {
         return this._isValueCreated;
     }
 
     public get value(): T {
-        if (!this._isValueCreated) {
-            this._value = this.valueFactory();
-            this._isValueCreated = true;
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._value = this.valueFactory();
+        this._isValueCreated = true;
+
+        if (this._valueLifetime) {
+            const reset = setTimeout(() => {
+                this._isValueCreated = false;
+                this._value = undefined;
+                clearTimeout(reset);
+            }, this._valueLifetime);
         }
 
         return this._value;
     }
 }
 
-export default Lazy;
+export class AsyncLazy<T> {
+    private _isValueCreated: boolean = false;
+    private _value: T | undefined;
+    private _valuePromise: Promise<T> | undefined;
+
+    public constructor(private readonly valueFactory: () => Promise<T>, private readonly _valueLifetime?: number) {
+    }
+
+    public async getValue(): Promise<T> {
+        if (this._isValueCreated) {
+            return this._value;
+        }
+
+        this._isValueCreated = false;
+
+        const isPrimaryPromise = this._valuePromise === undefined; // The first caller is "primary"
+        const fnLocalPromiseRef = this._valuePromise = this._valuePromise ?? this.valueFactory(); // Await any currently-running Promise if there is one
```

Comment:
If another execution changes `this._valuePromise` to `undefined` while this one is awaiting, I'm not totally sure what will happen. It might not be necessary.

Reply:
I am 99% sure `await this._valuePromise` will take the value of `this._valuePromise` at the start of await operation, so subsequent changes to that member will have no effect

Reply:
I'm 99% sure I agree. I'll test it out to cover the 1% uncertainty. :smile: